src/custom/constants/networks.ts

```diff
@@ -28,3 +38,142 @@ export const RPC_URLS: { [key in SupportedChainId]: string } = {
+    chainId: SupportedChainId.GOERLI,
+    name: 'goerli',
+  },
+  [SupportedChainId.GNOSIS_CHAIN]: {
+    chainId: SupportedChainId.GNOSIS_CHAIN,
+    name: 'gnosis',
+  },
+}
+
+/**
+ * We use this class for type inference.
+ *
+ * UrlJsonRpcProvider is an abstract class, therefore we have trouble inferring the type of its constructor.
+ * However, as all of the RPC providers we use extend UrlJsonRpcProvider, we can use this class to infer the type of the constructor, by turning it into a concrete one.
+ */
+class BaseUrlJsonRpcProvider extends UrlJsonRpcProvider {}
+type ProviderClass = Newable<typeof BaseUrlJsonRpcProvider>
+
+/**
+ * UrlJsonRpcProvider instances do not have a getSigner implementation, however we need it.
+ *
+ * As they are all based on JsonRpcProvider, we can use JsonRpcProvider's getSigner implementation.
+ *
+ * Why not use JsonRpcProvider directly? Because it does not have specific API provider optimisations,
+ * and it does not automatically support community resources.
+ *
+ * Why not instantiate and return a JsonRpcSigner? Because it has a constructor guard that prevents it from being instantiated directly.
+ * Intention with this guard is to prevent users from accidentally using a JsonRpcSigner with another provider. Which is not possible here.
+ */
+class InfuraProvider extends EthersInfuraProvider {}
+InfuraProvider.prototype.getSigner = JsonRpcProvider.prototype.getSigner
+class CloudflareProvider extends EthersCloudflareProvider {}
+CloudflareProvider.prototype.getSigner = JsonRpcProvider.prototype.getSigner
+class AlchemyProvider extends EthersAlchemyProvider {}
+AlchemyProvider.prototype.getSigner = JsonRpcProvider.prototype.getSigner
+class PocketProvider extends EthersPocketProvider {}
+PocketProvider.prototype.getSigner = JsonRpcProvider.prototype.getSigner
+class AnkrProvider extends EthersAnkrProvider {}
+AnkrProvider.prototype.getSigner = JsonRpcProvider.prototype.getSigner
+
+// List of public providers that don't require an API key, per chain support.
+const PUBLIC_PROVIDER_LIST: Record<SupportedChainId, ProviderClass[]> = {
+  [SupportedChainId.MAINNET]: [InfuraProvider, CloudflareProvider, AlchemyProvider, PocketProvider, AnkrProvider],
+  [SupportedChainId.GOERLI]: [InfuraProvider, AlchemyProvider, PocketProvider, AnkrProvider],
+  [SupportedChainId.GNOSIS_CHAIN]: [],
+}
+
+const RPC_TYPE: { [key in SupportedChainId]: 'infura' | 'custom' } = {
+  [SupportedChainId.MAINNET]: 'infura',
+  [SupportedChainId.GOERLI]: 'infura',
+  [SupportedChainId.GNOSIS_CHAIN]: 'custom',
+}
+
+function getProvider(chainId: SupportedChainId): JsonRpcProvider[] {
+  const result: JsonRpcProvider[] = []
+
+  if (window.ethereum) {
+    result.push(new Web3Provider(window.ethereum, RPC_CONFIG[chainId]))
+  }
+
+  result.pop()
+
+  if (RPC_TYPE[chainId] === 'infura') {
+    result.push(new InfuraProvider(RPC_CONFIG[chainId], INFURA_KEY))
+  } else {
+    result.push(new StaticJsonRpcProvider(RPC_URLS[chainId]))
+  }
```

Comment:
`RPC_URLS` contains infura urls for mainnet & goerli, and the default rpc for gnosis.

Does it make sense to use it here?
Intuitively, it would only ever be used for gnosis chain.

Might as well no longer export it and rename it to something more meaningful, like fallback rpc urls, static rpc urls, etc.

What do you think?

Reply:
It is a good point, although this was essentially a shot on my side for generalisation.

Simplified it - I think now we should be in a better shape.